source/adios2/operator/compress/CompressMGARD.cpp

```diff
@@ -139,28 +139,14 @@ size_t CompressMGARD::Operate(const char *dataIn, const Dims &blockStart,
         }
     }
 
-    size_t sizeOut = 0;
-    void *compressedData = nullptr;
+    // let mgard know the output buffer size
+    size_t sizeOut =
+        helper::GetTotalSize(blockCount, helper::GetDataTypeSize(type));
+    void *compressedData = bufferOut + bufferOutOffset;
     mgard_x::compress(mgardDim, mgardType, mgardCount, tolerance, s,
-                      errorBoundType, dataIn, compressedData, sizeOut, false);
-
-    if (bufferOutOffset + sizeOut >
-        helper::GetTotalSize(blockCount, helper::GetDataTypeSize(type)))
-    {
-        CompressNull c({});
-        bufferOutOffset =
-            c.Operate(dataIn, blockStart, blockCount, type, bufferOut);
-    }
-    else
-    {
-        std::memcpy(bufferOut + bufferOutOffset, compressedData, sizeOut);
-        bufferOutOffset += sizeOut;
-    }
+                      errorBoundType, dataIn, compressedData, sizeOut, true);
```

Comment:
The `true` flag at the end is telling the `compress` function that the output is pre-allocated?

Reply:
Yes. This is correct. MGARD should use the buffer provided by ADIOS for output. Internally, MGARD will automatically detect where the buffer is located (CPU or GPU) and invoke the correct memory copy function.